Create check_links.yml

erget opened the pull request:
Supports #132 .

erget commented:
MLC's pretty slow. Maybe https://github.com/marketplace/actions/link-checker ?

erget commented:
This is pretty nice and now the pipeline fails as expected. In the core it uses liche. From here the best thing will be to eliminate as many errors as possible and then [exclude wontfixes](https://github.com/raviqqe/liche). That's also a sustainable model moving forward if we know that certain files shouldn't be checked.

erget commented:
Status: I'm not so crazy about the current implementation. Currently I'm experimenting with [linkchecker](https://github.com/linkchecker/linkchecker) and considering using it in conjunction with [a cronjob generating an issue when broken links are found](https://github.com/lycheeverse/lychee-action#receiving-issues-containing-the-error-report).

Tried out something like this, now I've tried integrating it into a GitHub Action:

```bash
docker pull linkchecker/linkchecker
docker run --rm -it -u $(id -u):$(id -g) linkchecker/linkchecker http://cfconventions.org/ -o blacklist --check-extern  # Now that's more like it! 6m31s.
# wontfixes can be ignored with e.g. --ignore-url http://cfconventions.org/cf-conventions/cf-conventions.html 
```

Once I've got that running I want to fail the pipeline if it fails and put it into an issue if it produces output. Run periodically.

DocOtak commented:
This one looks pretty well maintained: https://github.com/lycheeverse/lychee-action and seems to be recommended by the "deprecated" https://github.com/marketplace/actions/link-checker

erget commented:
Yeah, I checked that out but it doesn't have recursion. In the end I built my own action from scratch that uses linkchecker to look at the actual website without scanning the sources. That works fine locally, but it never halts when I run it as a GitHub action - strangely, it keeps on reporting a heartbeat from the application, so it looks like it's not a straightforward environment issue.

I'll troubleshoot this more next week but in the meantime if you have a good idea I'm all ears :)

DocOtak commented:
Hm... recursively crawling the actual site over http sounds slow. Is there a reason to not just throw all the markdown files at lychee (via globbing).

Might want to kill that GH action... it has been going for at least 2 hours now.

erget commented:
@DocOtak I think I'm cursed. Would you have a few seconds to see what I'm doing wrong, or check my sanity at least?

I switched to Lychee - you're right, this does look better, and the checks complete much faster. Now the checks complete, and according to the documentation, the output path should be `./lychee/out.md`. However, adding a step that tries to preserve that as an artifact fails. I dug into the code of `lychee-action` and I don't understand why - it looks like the file should be there. 🤷 

If that works I believe I should be able to generate an issue off of the failed link checks, which would be a great first step. Although I'm not sure if the report will be sufficient, because `lychee-action` `cat`s that to `stdout` and it looks like it's not obvious where the failing links are located, but I'll cross that bridge when I can get my hands on the report in the first place!

Using the solution from last week (linkchecker, sadly not available as an action but with a Docker image) worked fine, executing in <10 minutes locally. It also ran on GitHub but never halted, I'm not sure what would cause that, which is why I abandoned it.

DocOtak commented:
Hey @erget sorry for the delay in response.
For reference, looks like the issue is related to https://github.com/lycheeverse/lychee-action/issues/1

However, I can see in one of your attempts you tried to access the tmp file directly... it's really strange that /tmp/lychee/out.md doesn't seem to exist....

Uh, Next thing I'd probably try is installing lychee myself via something like https://github.com/marketplace/actions/rust-cargo-install but I'm kinda at a loss on this one.

mre commented:
Sorry for the hiccup folks. It was a bug in the lychee action. Should be fixed now. Classic one-letter fix.
https://github.com/lycheeverse/lychee-action/commit/5abbdadaddae26c045f74832c946901b30d8566f
All credits go to @c-martinez for this one.
Let me know if you still run into problems. 😅

erget commented:
@mre beautiful, thank you very much! Works like it should.

sadielbartholomew commented on .github/workflows/check_links.yml:
```diff
@@ -0,0 +1,38 @@
+#
+# For more information on the actions used in this workflow, please see:
+#   https://github.com/actions/checkout
+#   https://github.com/marketplace/actions/build-jekyll
```
I can't see that Action (`build-jekyll`) here, is it used under-the-hood by another external action, or has it been left in by mistake from copying the `check_jekyll_build.yml` workflow as a template perhaps (I had a quick look at the earlier commits but there's quite a lot and I can't see anything)?

DocOtak replied:
I think in an earlier version of this action, it built the site and ran an HTML checker over it. This was switched to running a link checker directly over the source markdown files. So this can be probably be removed.

sadielbartholomew commented on .github/workflows/check_links.yml:
```diff
@@ -0,0 +1,38 @@
+
+jobs:
+  check_links:
+    name: Check Markdown links  # TODO: Check also Asciidoc
```
Is this TODO intended for a follow-on PR? Or should it be addressed pre-merge?

erget replied:
This is for a follow-on. Checking the Asciidoc would require a little bit more hands-on than I have time for ATM so I'll merge this with the MD checks to close the current issue. IMO it makes more sense to check the AsciiDoc where it lives in the other repo.

sadielbartholomew commented on .github/workflows/check_links.yml:
```diff
@@ -0,0 +1,38 @@
+#   https://github.com/actions/checkout
+#   https://github.com/marketplace/actions/build-jekyll
+#   https://github.com/actions/upload-artifact
+#
```
It would be good to add the new core external action, `https://github.com/lycheeverse/lychee-action`, to this list, as well as the relevant location of `peter-evans/create-issue-from-file@v2` which is also now used.

erget replied:
I just removed the "for further info" section, as it's more verbose than googling ;)

sadielbartholomew reviewed (commented):
I have a few questions and minor comments regarding the new workflow, but in general it looks sound and from inspecting at the logs of the final and succeeding 'Check Markdown links' job it has clearly identified lots of links across the site and determined that they are valid, so it seems to work! Great stuff.

I now just want to make sure it does indeed check the links work by putting in a dummy PR which (purely as a test) has a commit that breaks some links. If the workflow runs and fails with a 404 against those broken links, as I observe in the browser, I will be satisfied this works as it should and approve.

sadielbartholomew reviewed (commented):
Hi again @erget, just to follow up after my little experiment over on #141 as linked above (in which I added some comments to explain what I was doing and seeing): the workflow seems to work fine to detect broken links with the proviso that I had to make the small change:

```diff
git diff c88e44bca86c69f72a899952d25979141bc6696f
diff --git a/.github/workflows/check_links.yml b/.github/workflows/check_links.yml
index d10ff2e..c51d484 100644
--- a/.github/workflows/check_links.yml
+++ b/.github/workflows/check_links.yml
@@ -21,7 +21,7 @@ jobs:
       - name: Checkout
         uses: actions/checkout@v2
       - name: Check links
-        uses: lycheeverse/lychee-action@5abbdad
+        uses: lycheeverse/lychee-action@v1.0.4
         
       - name: Upload report
         if: ${{ always() }}

```
in response to an error that was raised (see https://github.com/cf-convention/cf-convention.github.io/pull/141#issuecomment-789461068).

Note I tested after the above change with success with using a named version rather than the full commit SHA corresponding to the commit currently referenced because I saw from the commit history of lycee that there was a new version released after that bug-fix commit. It is better to use a named version that a specific commit between versions, I think (though go with whatever you feel is most appropriate of course).

So I think that such a one-line change to either the named version or the full commit hash is required.

Otherwise, this is great. Please read my minor comments raised in the previous review set, but besides those and once the above is addressed, I am happy to approve this PR.

erget commented:
@sadielbartholomew thanks for https://github.com/cf-convention/cf-convention.github.io/pull/134#pullrequestreview-602558148 and the other comments - I definitely agree that a tagged reference is better than a sha, I had implemented that bit as a hotfix before a tag was available. I've added the updates you request and will squash and merge as soon as it runs through clean.
